examples/rsa/search_inference.py

```diff
@@ -129,18 +129,19 @@ def _weighted_mean(self, value, dim=0):
         for _ in range(value.dim() - 1):
             weights = weights.unsqueeze(-1)
         max_val = weights.max(dim)[0]
```

Comment:
Hmm the original code looks suspicious to me, in that we're calling `.max(dim)` but then `unsqueeze(-1)`. Shouldn't we instead `.max(dim=dim, keepdim=True)`? (also eliminating common subexpressions:)
```py
max_val = weights.max(dim, keepdim=True)[0]
probs = (weights - max_val).exp()
return (value * probs).sum(dim=dim) / probs.sum(dim=dim)
```
(BTW thanks for this fix, and sorry to bog down review with cleanup that's not entirely related to your PR!)

Reply:
Good call on `probs = ...`!

Due to pytorch's broadcasting that doesn't work, and I don't think `keepdim` has an effect. The original `unsqueeze()` calls are used to add `value.dim()-1` singleton dimensions to `weights`, ensuring that it aligns with `value` along dimension 0. Maybe it'd be a bit cleaner to write:

```python
weights = weights.reshape([-1] + (value.dim() - 1) * [1])
```

?

Reply:
I think the unsqueezes or reshapes are needed, but the remaining can be changed as per @fritzo's suggestion (which you already did, I see). 

> I am also curious whether there's a reason for _weighted_mean() to have the dim argument.

I think it was put in to support the possibility of multiple chains, with `dim=0` corresponds to the chain, but it will likely need other fixes to work correctly.

Reply:
Thanks for explaining!

Reply:
Ok, I've pushed a version using `reshape()`. `_weighted_mean()` will break if `dim != 0`, but I guess I'll leave it as is given @neerajprad's comment.